Fix #105, #107, #125, #134 and #209

sunghee-hwang opened the pull request:
<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***

### :boom: Error: 400 Bad Request :boom: ###

[PR Preview](https://github.com/tobie/pr-preview#pr-preview) failed to build. _(Last tried on Jan 10, 2023, 12:30 AM UTC)_.

<details>
<summary>More</summary>


PR Preview relies on a number of web services to run. There seems to be an issue with the following one:

:rotating_light: [CSS Spec Preprocessor](https://api.csswg.org/bikeshed/) - CSS Spec Preprocessor is the web service used to build Bikeshed specs.

:link: [Related URL](https://api.csswg.org/bikeshed/?url=https%3A%2F%2Fraw.githubusercontent.com%2FAOMediaCodec%2Fiamf%2F1fe95f707cd51cba19f745a0fdd07d7fb6c27d96%2Findex.bs&md-warning=not%20ready)

```
Error running preprocessor, returned code: 2.
Your document appears to use spaces to indent, but line 372 starts with tabs.
Your document appears to use spaces to indent, but line 373 starts with tabs.
Your document appears to use spaces to indent, but line 374 starts with tabs.
Your document appears to use spaces to indent, but line 375 starts with tabs.
Your document appears to use spaces to indent, but line 376 starts with tabs.
Your document appears to use spaces to indent, but line 377 starts with tabs.
Your document appears to use spaces to indent, but line 378 starts with tabs.
Your document appears to use spaces to indent, but line 379 starts with tabs.
Your document appears to use spaces to indent, but line 380 starts with tabs.
Your document appears to use spaces to indent, but line 381 starts with tabs.
Your document appears to use spaces to indent, but line 382 starts with tabs.
Your document appears to use spaces to indent, but line 384 starts with tabs.
Your document appears to use spaces to indent, but line 385 starts with tabs.
Your document appears to use spaces to indent, but line 908 starts with tabs.
Your document appears to use spaces to indent, but line 909 starts with tabs.
Your document appears to use spaces to indent, but line 910 starts with tabs.
Your document appears to use spaces to indent, but line 1680 starts with tabs.
Your document appears to use spaces to indent, but line 1681 starts with tabs.
Your document appears to use spaces to indent, but line 1682 starts with tabs.
Your document appears to use spaces to indent, but line 1683 starts with tabs.
Your document appears to use spaces to indent, but line 1684 starts with tabs.
Your document appears to use spaces to indent, but line 1685 starts with tabs.
Your document appears to use spaces to indent, but line 1736 starts with tabs.
Your document appears to use spaces to indent, but line 1737 starts with tabs.
Your document appears to use spaces to indent, but line 1761 starts with tabs.
Your document appears to use spaces to indent, but line 1762 starts with tabs.
Your document appears to use spaces to indent, but line 1763 starts with tabs.
Your document appears to use spaces to indent, but line 1969 starts with tabs.
Your document appears to use spaces to indent, but line 1970 starts with tabs.
Your document appears to use spaces to indent, but line 1971 starts with tabs.
Your document appears to use spaces to indent, but line 1975 starts with tabs.
Your document appears to use spaces to indent, but line 1976 starts with tabs.
Your document appears to use spaces to indent, but line 1977 starts with tabs.
Your document appears to use spaces to indent, but line 1979 starts with tabs.
Your document appears to use spaces to indent, but line 1980 starts with tabs.
Your document appears to use spaces to indent, but line 1981 starts with tabs.
Your document appears to use spaces to indent, but line 1982 starts with tabs.
Your document appears to use spaces to indent, but line 2125 starts with tabs.
Your document appears to use spaces to indent, but line 2126 starts with tabs.
Your document appears to use spaces to indent, but line 2128 starts with tabs.
Your document appears to use spaces to indent, but line 2129 starts with tabs.
Your document appears to use spaces to indent, but line 2160 starts with tabs.
Your document appears to use spaces to indent, but line 2161 starts with tabs.
Your document appears to use spaces to indent, but line 2162 starts with tabs.
Your document appears to use spaces to indent, but line 2163 starts with tabs.
Your document appears to use spaces to indent, but line 2166 starts with tabs.
Your document appears to use spaces to indent, but line 2167 starts with tabs.
Your document appears to use spaces to indent, but line 2168 starts with tabs.
Your document appears to use spaces to indent, but line 2173 starts with tabs.
Your document appears to use spaces to indent, but line 2174 starts with tabs.
Your document appears to use spaces to indent, but line 2190 starts with tabs.
Your document appears to use spaces to indent, but line 2202 starts with tabs.
Your document appears to use spaces to indent, but line 2203 starts with tabs.
Your document appears to use spaces to indent, but line 2205 starts with tabs.
Your document appears to use spaces to indent, but line 2206 starts with tabs.
Your document appears to use spaces to indent, but line 2207 starts with tabs.
Your document appears to use spaces to indent, but line 2217 starts with tabs.
Your document appears to use spaces to indent, but line 2237 starts with tabs.
Your document appears to use spaces to indent, but line 2239 starts with tabs.
Your document appears to use spaces to indent, but line 2240 starts with tabs.
Your document appears to use spaces to indent, but line 2241 starts with tabs.
Your document appears to use spaces to indent, but line 2242 starts with tabs.
Your document appears to use spaces to indent, but line 2244 starts with tabs.
Your document appears to use spaces to indent, but line 2245 starts with tabs.
Your document appears to use spaces to indent, but line 2246 starts with tabs.
Your document appears to use spaces to indent, but line 2270 starts with tabs.
Your document appears to use spaces to indent, but line 2271 starts with tabs.
Your document appears to use spaces to indent, but line 2272 starts with tabs.
Your document appears to use spaces to indent, but line 2273 starts with tabs.
Your document appears to use spaces to indent, but line 2275 starts with tabs.
Your document appears to use spaces to indent, but line 2276 starts with tabs.
Your document appears to use spaces to indent, but line 2325 starts with tabs.
Your document appears to use spaces to indent, but line 2326 starts with tabs.
Your document appears to use spaces to indent, but line 2327 starts with tabs.
Your document appears to use spaces to indent, but line 2328 starts with tabs.
Your document appears to use spaces to indent, but line 2329 starts with tabs.
Your document appears to use spaces to indent, but line 2330 starts with tabs.
Your document appears to use spaces to indent, but line 2331 starts with tabs.
Your document appears to use spaces to indent, but line 2354 starts with tabs.
Your document appears to use spaces to indent, but line 2355 starts with tabs.
Your document appears to use spaces to indent, but line 2356 starts with tabs.
Your document appears to use spaces to indent, but line 2357 starts with tabs.
Your document appears to use spaces to indent, but line 2358 starts with tabs.
Your document appears to use spaces to indent, but line 2359 starts with tabs.
Your document appears to use spaces to indent, but line 2367 starts with tabs.
Your document appears to use spaces to indent, but line 2368 starts with tabs.
Your document appears to use spaces to indent, but line 2369 starts with tabs.
Your document appears to use spaces to indent, but line 2372 starts with tabs.
Your document appears to use spaces to indent, but line 2373 starts with tabs.
Your document appears to use spaces to indent, but line 2375 starts with tabs.
Your document appears to use spaces to indent, but line 2376 starts with tabs.
Your document appears to use spaces to indent, but line 2377 starts with tabs.
Your document appears to use spaces to indent, but line 2378 starts with tabs.
Your document appears to use spaces to indent, but line 2384 starts with tabs.
Your document appears to use spaces to indent, but line 2385 starts with tabs.
Your document appears to use spaces to indent, but line 2386 starts with tabs.
Your document appears to use spaces to indent, but line 2522 starts with tabs.
Your document appears to use spaces to indent, but line 2543 starts with tabs.
Your document appears to use spaces to indent, but line 2544 starts with tabs.
Your document appears to use spaces to indent, but line 2545 starts with tabs.
Your document appears to use spaces to indent, but line 2546 starts with tabs.
Your document appears to use spaces to indent, but line 2547 starts with tabs.
Your document appears to use spaces to indent, but line 2548 starts with tabs.
Your document appears to use spaces to indent, but line 2549 starts with tabs.
Your document appears to use spaces to indent, but line 2551 starts with tabs.
Your document appears to use spaces to indent, but line 2552 starts with tabs.
Your document appears to use spaces to indent, but line 2553 starts with tabs.
Your document appears to use spaces to indent, but line 2554 starts with tabs.
Your document appears to use spaces to indent, but line 2556 starts with tabs.
Your document appears to use spaces to indent, but line 2564 starts with tabs.
Your document appears to use spaces to indent, but line 2565 starts with tabs.
Your document appears to use spaces to indent, but line 2566 starts with tabs.
Your document appears to use spaces to indent, but line 2567 starts with tabs.
Your document appears to use spaces to indent, but line 2568 starts with tabs.
Your document appears to use spaces to indent, but line 2569 starts with tabs.
Your document appears to use spaces to indent, but line 2570 starts with tabs.
Your document appears to use spaces to indent, but line 2571 starts with tabs.
Your document appears to use spaces to indent, but line 2572 starts with tabs.
Your document appears to use spaces to indent, but line 2578 starts with tabs.
Your document appears to use spaces to indent, but line 2579 starts with tabs.
Your document appears to use spaces to indent, but line 2580 starts with tabs.
Your document appears to use spaces to indent, but line 2581 starts with tabs.
Your document appears to use spaces to indent, but line 2582 starts with tabs.
Your document appears to use spaces to indent, but line 2583 starts with tabs.
Your document appears to use spaces to indent, but line 2586 starts with tabs.
Your document appears to use spaces to indent, but line 2587 starts with tabs.
Your document appears to use spaces to indent, but line 2593 starts with tabs.
Your document appears to use spaces to indent, but line 2594 starts with tabs.
Your document appears to use spaces to indent, but line 2595 starts with tabs.
Your document appears to use spaces to indent, but line 2596 starts with tabs.
Your document appears to use spaces to indent, but line 2597 starts with tabs.
Your document appears to use spaces to indent, but line 2598 starts with tabs.
Your document appears to use spaces to indent, but line 2599 starts with tabs.
Your document appears to use spaces to indent, but line 2600 starts with tabs.
Your document appears to use spaces to indent, but line 2601 starts with tabs.
Your document appears to use spaces to indent, but line 2602 starts with tabs.
Your document appears to use spaces to indent, but line 2603 starts with tabs.
Your document appears to use spaces to indent, but line 2604 starts with tabs.
Your document appears to use spaces to indent, but line 2605 starts with tabs.
Your document appears to use spaces to indent, but line 2606 starts with tabs.
Your document appears to use spaces to indent, but line 2607 starts with tabs.
Your document appears to use spaces to indent, but line 2608 starts with tabs.
Your document appears to use spaces to indent, but line 2609 starts with tabs.
Your document appears to use spaces to indent, but line 2610 starts with tabs.
Your document appears to use spaces to indent, but line 2611 starts with tabs.
Your document appears to use spaces to indent, but line 2612 starts with tabs.
Your document appears to use spaces to indent, but line 2613 starts with tabs.
Your document appears to use spaces to indent, but line 2614 starts with tabs.
Your document appears to use spaces to indent, but line 2615 starts with tabs.
Your document appears to use spaces to indent, but line 2616 starts with tabs.
Your document appears to use spaces to indent, but line 2617 starts with tabs.
Your document appears to use spaces to indent, but line 2619 starts with tabs.
Your document appears to use spaces to indent, but line 2620 starts with tabs.
Your document appears to use spaces to indent, but line 2625 starts with tabs.
Your document appears to use spaces to indent, but line 2628 starts with tabs.
Your document appears to use spaces to indent, but line 2631 starts with tabs.
Your document appears to use spaces to indent, but line 2635 starts with tabs.
Your document appears to use spaces to indent, but line 2636 starts with tabs.
Your document appears to use spaces to indent, but line 2655 starts with tabs.
Your document appears to use spaces to indent, but line 2656 starts with tabs.
Your document appears to use spaces to indent, but line 2657 starts with tabs.
Your document appears to use spaces to indent, but line 2658 starts with tabs.
Your document appears to use spaces to indent, but line 2659 starts with tabs.
Your document appears to use spaces to indent, but line 2660 starts with tabs.
Your document appears to use spaces to indent, but line 2661 starts with tabs.
Your document appears to use spaces to indent, but line 2663 starts with tabs.
Your document appears to use spaces to indent, but line 2759 starts with tabs.
Your document appears to use spaces to indent, but line 2760 starts with tabs.
Your document appears to use spaces to indent, but line 2761 starts with tabs.
Your document appears to use spaces to indent, but line 2762 starts with tabs.
Your document appears to use spaces to indent, but line 2763 starts with tabs.
Your document appears to use spaces to indent, but line 2764 starts with tabs.
Your document appears to use spaces to indent, but line 2765 starts with tabs.
Your document appears to use spaces to indent, but line 2766 starts with tabs.
Your document appears to use spaces to indent, but line 2767 starts with tabs.
Your document appears to use spaces to indent, but line 2828 starts with tabs.
Your document appears to use spaces to indent, but line 2829 starts with tabs.
Your document appears to use spaces to indent, but line 2831 starts with tabs.
Your document appears to use spaces to indent, but line 2833 starts with tabs.
Your document appears to use spaces to indent, but line 2834 starts with tabs.
Your document appears to use spaces to indent, but line 2835 starts with tabs.
Your document appears to use spaces to indent, but line 2839 starts with tabs.
Your document appears to use spaces to indent, but line 2840 starts with tabs.
Your document appears to use spaces to indent, but line 2842 starts with tabs.
Your document appears to use spaces to indent, but line 2854 starts with tabs.
Your document appears to use spaces to indent, but line 2856 starts with tabs.
Your document appears to use spaces to indent, but line 2858 starts with tabs.
Your document appears to use spaces to indent, but line 2859 starts with tabs.
Your document appears to use spaces to indent, but line 2860 starts with tabs.
Your document appears to use spaces to indent, but line 2865 starts with tabs.
WARNING: The heading 'Annex A: Audio Substream Gaps' needs a manually-specified ID.
WARNING: The heading 'A gap within an audio substream' needs a manually-specified ID.
WARNING: The heading 'A gap due to packet loss' needs a manually-specified ID.
WARNING: The heading 'A gap between two audio substreams' needs a manually-specified ID.
WARNING: Multiple elements have the same ID 'iacgeneration-postprocessing-drc'.
Deduping, but this ID may not be stable across revisions.
WARNING: At least one &lt;img> doesn't have its size set (&lt;img src="images/Conceptual%20IAC%20Architecture.png">), but given the type of input document, Bikeshed can't figure out what the size should be.
Either set 'width'/'height' manually, or opt out of auto-detection by setting the 'no-autosize' attribute.
WARNING: Multiple elements have the same ID 'control_point_relative_time'.
Deduping, but this ID may not be stable across revisions.
FATAL ERROR: Multiple local 'dfn' &lt;dfn>s have the same linking text 'control_point_relative_time'.
 ✘  Did not generate, due to fatal errors
```

_If you don't have enough information above to solve the error by yourself (or to understand to which web service the error is related to, if any), please [file an issue](https://github.com/tobie/pr-preview/issues/new?title=Error%20not%20surfaced%20properly&body=See%20AOMediaCodec/iamf%23215.)._
</details>

sunghee-hwang commented:
Here are major changes
- 4.1.8 Mix Presentation OBU Syntax: reformulated under the umbrella of sub-mixes.
- 4.2.1 Added Binaural
- 4.2.6 Prepared spec text for render_config().
- 5. Profiles: added restrictions
- 9.3.1 Selection a Mix Presentation
- 9.3.2 Rendering an Audio Element
- 9.4.1 Loudness Normalization
- 9.4.2 DRC Control (Deleted)
- 9.5.1 Static Down-mix Matrix (Deleted)
- 10.2.6 Mix Presentation Encoding
- 10.2.6.1 Element Mix Config

sunghee-hwang commented on index.bs:
```diff
@@ -971,12 +1080,15 @@ The IA decoder shall select one of one or more channel audios provided by scalab
 <dfn noexport>num_layers</dfn> shall indicate the number of ChannelGroups for scalable channel audio. It shall not be set to zero and its maximum number shall be limited to 6.
+- For Binaural, this field shall be set to 1.
+
+ISUUE: Added for Binaural
```
Logically, stereo is possible to be scalable, 1ch/2ch.

felicialim replied:
Oh, I meant the "ISSUE" line

sunghee-hwang replied:
Oh I see.

sunghee-hwang commented on index.bs:
```diff
@@ -631,71 +668,143 @@ parameter_name : Parameter name.
+      element_mix_config();
+    }
+    output_mix_config();
+    layout measured_layout;
```
Update it.

sunghee-hwang commented on index.bs:
```diff
@@ -2056,41 +2317,130 @@ Recommend values for specific codecs are as follows
-Loudness normalization is done by adjusting a loudness level to -24 LKFS based on the loudness value of the target channel layout (i.e. CL #i) which is signaled in [=Channel_Audio_Layer_Config()=] or the loudness value in mix presentation OBU.
+Loudness normalization should be done by adjusting the loudness level to a target value, using the integrated loudness and true peak information provided in loudness_info(). If the true peak information is not available, the digital peak information should be used.
+
+The rendered layout that was used to measure the loudness information of a sub-mix is provided by measured_layout or as [=loudspeaker_layout=] in channel_audio_layer_config(). 
```
Updated it

sunghee-hwang commented on index.bs:
```diff
@@ -1272,87 +1401,214 @@ class mix_target_layout(mix_target_layout_type) {
+      signed int (16) azimuth_min;
+      signed int (16) elevation_max;
+      signed int (16) elevation_min;
+      signed int (16) distance_max;
```
I fixed Q8.7 to Q7.8 too.

felicialim replied:
Actually I think we want Q8.7 here. This allows up to 1-degree resolution on the integer side. 7 bits for the decimal part is probably allowing more precision than needed, but using 16 bits allows for byte alignment.

sunghee-hwang replied:
Ops, sorry for that. I reverted it.

tdaede commented:
Let's review this and plan to merge by next week.

felicialim commented on index.bs:
```diff
@@ -795,7 +795,7 @@ class loudness_info() {
  Bit : Type of information provided
   0  : True peak
- 2~8 : Reserved
+ 1~8 : Reserved
```
The upper limit should be 7 I think? I.e. 1~7

sunghee-hwang commented:
Updated to implement multiple loudness information. So, I removed loudness information in scalable_channel_audio_layer instead.
Based on this I updated selecting rule of mix presentations and mix presentation encoding as well.

felicialim commented on index.bs:
```diff
@@ -666,71 +703,153 @@ A ChannelGroup is defined in [[#iacgeneration]]. The order of the substreams in
+
+<dfn noexport>num_layouts</dfn> specifies the number of layouts for this sub-mix which the loudness informations were measured on.
+
+<dfn noexport>measured_layout_for_loudness</dfn> identifies the layout that was used to measure the loudness information provided in this sub-mix.
```
nit: actually, maybe a better name is loudness_layout (or if you have another suggestion). Sorry for the churn on this

felicialim commented on index.bs:
```diff
@@ -2064,146 +2330,133 @@ Recommend values for specific codecs are as follows
+2. If there are more than one valid mixes remaining, the IA parser should select an appropriate mix for rendering, in the following order.
+	1. If the playback layout is binaural, i.e. headphones:
+		1. Select the mix with [=audio_element_id=] whose [=loudspeaker_layout=] is BINAURAL.
+		2. If there is no such mix, select the mix with the closest available [=measured_layout_for_loudness=].
```
Should we make it more explicit what "closest available" is? A proposal: select the mix with measured_layout_for_loudness equal to BINAURAL, otherwise select the mix with the highest available measured_layout_for_loudness.

sunghee-hwang replied:
It makes a sense.

felicialim commented on index.bs:
```diff
@@ -2064,146 +2330,133 @@ Recommend values for specific codecs are as follows
+		2. If there is no such mix, select the mix with the closest available [=measured_layout_for_loudness=].
+	2. If the playback layout is loudspeakers:
+		1. If there is a mix with an [=measured_layout_for_loudness=] that matches the playback loudspeaker layout, it should be selected. If there are more than one matching mixes, the first one should be selected.
+		2. If there is no such mix, select the mix presentation with the closest available [=measured_layout_for_loudness=].
```
same comment here about making "closest available" more explicit

tdaede commented:
needs rebase but can be merged
